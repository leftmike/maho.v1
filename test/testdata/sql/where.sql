--
-- Test SELECT WHERE with indexes
--
-- {{Sort .Global false}}

DROP TABLE IF EXISTS tbl1;

CREATE TABLE tbl1 (
    c1 int primary key,
    c2 int,
    c3 int,
    c4 int,
    c5 int
);

CREATE INDEX idx1 ON tbl1 (c2 DESC);

CREATE INDEX idx2 ON tbl1 (c3, c4);

CREATE UNIQUE INDEX idx3 ON tbl1 (c5);

INSERT INTO tbl1 (c1, c2, c3, c4, c5) VALUES
    (1, 10, 100, -1, -1),
    (2, 10, 100, -2, -2),
    (3, 20, 100, -2, -3),
    (4, 20, 200, -3, -4),
    (5, 20, 200, -4, -5),
    (6, 30, 300, -5, -6),
    (7, 40, 300, -5, -7),
    (8, 40, 300, -5, -8),
    (9, 40, 300, -6, -9),
    (10, 50, 400, -6, -10),
    (11, 60, 500, -6, -11),
    (12, 70, 600, -6, -12),
    (13, 80, 700, -7, -13);

SELECT * FROM tbl1;

SELECT * FROM tbl1@idx1;

SELECT * FROM tbl1@idx2;

SELECT * FROM tbl1@idx3;

EXPLAIN SELECT * FROM tbl1 WHERE c1 = 5;

EXPLAIN SELECT * FROM tbl1@idx1 WHERE c2 = 5;
