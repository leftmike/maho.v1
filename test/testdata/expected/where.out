--
-- Test SELECT WHERE with indexes
--
-- {{Sort .Global false}}
DROP TABLE IF EXISTS tbl1;
CREATE TABLE tbl1 (
    c1 int primary key,
    c2 int,
    c3 int,
    c4 int,
    c5 int
);
CREATE INDEX idx1 ON tbl1 (c2 DESC);
CREATE INDEX idx2 ON tbl1 (c3, c4);
CREATE UNIQUE INDEX idx3 ON tbl1 (c5);
INSERT INTO tbl1 (c1, c2, c3, c4, c5) VALUES
    (1, 10, 100, -1, -1),
    (2, 10, 100, -2, -2),
    (3, 20, 100, -2, -3),
    (4, 20, 200, -3, -4),
    (5, 20, 200, -4, -5),
    (6, 30, 300, -5, -6),
    (7, 40, 300, -5, -7),
    (8, 40, 300, -5, -8),
    (9, 40, 300, -6, -9),
    (10, 50, 400, -6, -10),
    (11, 60, 500, -6, -11),
    (12, 70, 600, -6, -12),
    (13, 80, 700, -7, -13);
SELECT * FROM tbl1;
    c1 c2  c3 c4  c5
    -- --  -- --  --
  1  1 10 100 -1  -1
  2  2 10 100 -2  -2
  3  3 20 100 -2  -3
  4  4 20 200 -3  -4
  5  5 20 200 -4  -5
  6  6 30 300 -5  -6
  7  7 40 300 -5  -7
  8  8 40 300 -5  -8
  9  9 40 300 -6  -9
 10 10 50 400 -6 -10
 11 11 60 500 -6 -11
 12 12 70 600 -6 -12
 13 13 80 700 -7 -13
(13 rows)
SELECT * FROM tbl1@idx1;
    c2 c1
    -- --
  1 80 13
  2 70 12
  3 60 11
  4 50 10
  5 40  7
  6 40  8
  7 40  9
  8 30  6
  9 20  3
 10 20  4
 11 20  5
 12 10  1
 13 10  2
(13 rows)
SELECT * FROM tbl1@idx2;
     c3 c4 c1
     -- -- --
  1 100 -2  2
  2 100 -2  3
  3 100 -1  1
  4 200 -4  5
  5 200 -3  4
  6 300 -6  9
  7 300 -5  6
  8 300 -5  7
  9 300 -5  8
 10 400 -6 10
 11 500 -6 11
 12 600 -6 12
 13 700 -7 13
(13 rows)
SELECT * FROM tbl1@idx3;
     c5 c1
     -- --
  1 -13 13
  2 -12 12
  3 -11 11
  4 -10 10
  5  -9  9
  6  -8  8
  7  -7  7
  8  -6  6
  9  -5  5
 10  -4  4
 11  -3  3
 12  -2  2
 13  -1  1
(13 rows)
SET pushdown_where = true;
SHOW pushdown_where;
   pushdown_where
   --------------
 1           true
(1 row)
EXPLAIN SELECT * FROM tbl1 WHERE c1 = 5;
              tree  field      description
              ----  -----      -----------
 1          select                        
 2  +-- scan table                        
 3               |  table test.public.tbl1
 4               | filter           c1 = 5
(4 rows)
SELECT * FROM tbl1 WHERE c1 = 5;
   c1 c2  c3 c4 c5
   -- --  -- -- --
 1  5 20 200 -4 -5
(1 row)
EXPLAIN SELECT * FROM tbl1@idx1 WHERE c2 = 50;
                   tree field      description
                   ---- -----      -----------
 1               select                       
 2           +-- filter                       
 3                    |  expr     "=="(c2, 50)
 4       +-- scan index                       
 5                    | table test.public.tbl1
 6                    | index             idx1
(6 rows)
SELECT * FROM tbl1@idx1 WHERE c2 = 50;
   c2 c1
   -- --
 1 50 10
(1 row)
SELECT * FROM tbl1@idx1 WHERE c2 = 40;
   c2 c1
   -- --
 1 40  7
 2 40  8
 3 40  9
(3 rows)
SELECT * FROM tbl1@idx2 WHERE c4 = -5 AND c3 = 300;
    c3 c4 c1
    -- -- --
 1 300 -5  6
 2 300 -5  7
 3 300 -5  8
(3 rows)
SELECT * FROM tbl1@idx2 WHERE c3 = 300 AND c4 = -6;
    c3 c4 c1
    -- -- --
 1 300 -6  9
(1 row)
SELECT * FROM tbl1@idx3 WHERE c5 = -11;
    c5 c1
    -- --
 1 -11 11
(1 row)
SET pushdown_where = false;
SHOW pushdown_where;
   pushdown_where
   --------------
 1          false
(1 row)
EXPLAIN SELECT * FROM tbl1 WHERE c1 = 5;
                   tree field      description
                   ---- -----      -----------
 1               select                       
 2           +-- filter                       
 3                    |  expr      "=="(c1, 5)
 4       +-- scan table                       
 5                    | table test.public.tbl1
(5 rows)
SELECT * FROM tbl1 WHERE c1 = 5;
   c1 c2  c3 c4 c5
   -- --  -- -- --
 1  5 20 200 -4 -5
(1 row)
EXPLAIN SELECT * FROM tbl1@idx1 WHERE c2 = 50;
                   tree field      description
                   ---- -----      -----------
 1               select                       
 2           +-- filter                       
 3                    |  expr     "=="(c2, 50)
 4       +-- scan index                       
 5                    | table test.public.tbl1
 6                    | index             idx1
(6 rows)
SELECT * FROM tbl1@idx1 WHERE c2 = 50;
   c2 c1
   -- --
 1 50 10
(1 row)
SELECT * FROM tbl1@idx1 WHERE c2 = 40;
   c2 c1
   -- --
 1 40  7
 2 40  8
 3 40  9
(3 rows)
SELECT * FROM tbl1@idx2 WHERE c4 = -5 AND c3 = 300;
    c3 c4 c1
    -- -- --
 1 300 -5  6
 2 300 -5  7
 3 300 -5  8
(3 rows)
SELECT * FROM tbl1@idx2 WHERE c3 = 300 AND c4 = -6;
    c3 c4 c1
    -- -- --
 1 300 -6  9
(1 row)
SELECT * FROM tbl1@idx3 WHERE c5 = -11;
    c5 c1
    -- --
 1 -11 11
(1 row)
